onnxruntime/core/mlas/lib/qladd.cpp

```diff
@@ -533,17 +532,93 @@ MlasQLinearAddKernelHelper(
         // Pack, saturate, and add output zero point.
         const int16x8_t vacc0 = vqaddq_s16(vcombine_s16(vqmovn_s32(vacc0_lo), vqmovn_s32(vacc0_hi)), VectorZeroPointC);
         const int16x8_t vacc1 = vqaddq_s16(vcombine_s16(vqmovn_s32(vacc1_lo), vqmovn_s32(vacc1_hi)), VectorZeroPointC);
-        vc = SUI::combine_i8_s16(vacc0, vacc1);
+        typename SUI::i8x16_t vc = SUI::combine_i8_s16(vacc0, vacc1);
 
         n -= 16;
-        if (n < 0) break;
-
         SUI::vst1q_i8(OutputC, vc);
         OutputC += 16;
     }
 
-    if (n < 0) {
-        n += 16;
+    if (n > 0) {
+        typename SUI::T TailDataA[16] = { 0 };
+        typename SUI::T TailDataB[16] = { 0 };
+
+        if (!IsScalarA) {
+            typename SUI::T* tail = TailDataA;
+            typename SUI::T* tail_end = tail + n;
+            do {
+                *tail = *(const typename SUI::T*)InputA;
+                tail++;
+                InputA++;
+            } while (tail < tail_end);
+        }
+        if (!IsScalarB) {
+            typename SUI::T* tail = TailDataB;
+            typename SUI::T* tail_end = tail + n;
+            do {
+                *tail = *(const typename SUI::T*)InputB;
+                tail++;
+                InputB++;
+            } while (tail < tail_end);
+        }
+
```

Comment:
Would memcpy be a better choice?

Reply:
Could be. Yet we have some worry on building system, so try not link outside functions.

Reply:
memcpy is standard highly optimized compiler intrinsic (inline) function. Plus, for small amount of data, compilers replace it with MOV instructions.

---
In reply to: [451846218](https://github.com/microsoft/onnxruntime/pull/4455#discussion_r451846218) [](ancestors = 451846218)